src/hotspot/share/opto/addnode.cpp

```diff
@@ -395,9 +395,178 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
     }
   }
 
+  // Convert a + a + ... + a into a*n
+  Node* serial_additions = convert_serial_additions(phase, can_reshape, bt);
+  if (serial_additions != nullptr) {
+    return serial_additions;
+  }
+
   return AddNode::Ideal(phase, can_reshape);
 }
 
+// Convert a + a + ... + a into a*n
+Node* AddNode::convert_serial_additions(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+  if (find_power_of_two_addition_pattern(this, bt, nullptr) != nullptr) {
```

Comment:
Can you a comment that explain the need for this (what you replied in the PR comment essentially)?